moveit_ros/perception/handeye_calibration_target/src/handeye_aruco_target.cpp

```diff
@@ -0,0 +1,258 @@
+    cv::Ptr<cv::aruco::GridBoard> board =
+        cv::aruco::GridBoard::create(markers_x_, markers_y_, float(marker_size_), float(separation_), dictionary);
+
+    // Create target image
+    board->draw(image_size, image, separation_, border_bits_);
+  }
+  catch (cv::Exception& e)
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Aruco target image creation exception: " << e.what());
+    return false;
+  }
+
+  return true;
+}
+
+bool HandEyeArucoTarget::detectTargetPose(cv::Mat& image)
+{
+  std::lock_guard<std::mutex> lck(base_mtx_);
+  try
+  {
+    // Detect aruco board
+    aruco_mtx_.lock();
+    cv::Ptr<cv::aruco::Dictionary> dictionary = cv::aruco::getPredefinedDictionary(dict_);
+    cv::Ptr<cv::aruco::GridBoard> board =
+        cv::aruco::GridBoard::create(markers_x_, markers_y_, marker_size_real_, marker_seperation_real_, dictionary);
+    aruco_mtx_.unlock();
+    cv::Ptr<cv::aruco::DetectorParameters> params_ptr(new cv::aruco::DetectorParameters());
+#if CV_MAJOR_VERSION == 3 && CV_MINOR_VERSION == 2
+    params_ptr->doCornerRefinement = true;
+#else
+    params_ptr->cornerRefinementMethod = cv::aruco::CORNER_REFINE_NONE;
+#endif
+
+    std::vector<int> ids;
+    std::vector<std::vector<cv::Point2f>> corners;
+    cv::aruco::detectMarkers(image, dictionary, corners, ids, params_ptr);
+    if (ids.empty())
+    {
+      ROS_DEBUG_STREAM_NAMED(LOGNAME, "No aruco marker detected.");
+      return false;
+    }
+
+    // Refine markers borders
+    std::vector<std::vector<cv::Point2f>> rejected_corners;
+    cv::aruco::refineDetectedMarkers(image, board, corners, ids, rejected_corners, camera_matrix_, distor_coeffs_);
+
+    // Estimate aruco board pose
+    int valid = cv::aruco::estimatePoseBoard(corners, ids, board, camera_matrix_, distor_coeffs_, rvect_, tvect_);
+
+    // Draw the markers and frame axis if at least one marker is detected
+    if (valid == 0)
+    {
+      ROS_DEBUG_STREAM_NAMED(LOGNAME, "Cannot estimate aruco target pose.");
+      return false;
+    }
+
+    if (std::log10(std::fabs(rvect_[0])) > 10 || std::log10(std::fabs(rvect_[0])) > 10 ||
+        std::log10(std::fabs(rvect_[0])) > 10 || std::log10(std::fabs(tvect_[0])) > 10 ||
+        std::log10(std::fabs(tvect_[0])) > 10 || std::log10(std::fabs(tvect_[0])) > 10)
+    {
+      ROS_ERROR_STREAM_NAMED(LOGNAME, "Invalid target pose, please check CameraInfo msg.");
+    }
+
+    cv::Mat image_color;
+    cv::cvtColor(image, image_color, cv::COLOR_GRAY2RGB);
+    cv::aruco::drawDetectedMarkers(image_color, corners);
+    cv::aruco::drawAxis(image_color, camera_matrix_, distor_coeffs_, rvect_, tvect_, 0.1);
+    image = image_color;
+  }
+  catch (const cv::Exception& e)
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Aruco target detection exception: " << e.what());
+    return false;
+  }
+
+  return true;
+}
+
+geometry_msgs::TransformStamped HandEyeArucoTarget::getPose(std::string& frame_id)
```

Comment:
These three functions `getPose`, `getTFQuaternion`, `getTranslationVect` silently use the member variables/state, but I wouldn't have guessed it from the names. I would either change the functions to take the current state as input (convert from `rvect_`/`tvect_` to TF) or rename the functions to `getCurrentPose` etc., to make it clear that they use internal data.

It's also peculiar that `getPose` returns a `TransformStamped` and not a `Pose`.

Reply:
Reasonable. `getTFQuaternion` and `getTranslationVect` were changed to `convertToTFQuaternion` and `convertToStdVector`. And they take current state `rvect_/tvect_` as input arguments. I hope this would make them more clear that they are just serving as state converter.

`getPose` was changed to `getTransformStamped`. And some introduction of this function was put into the header file of plugin base `handeye_target_base.h`